### PR TITLE
[feature](function) support hll functions hll_from_base64, hll_to_base64

### PR DESCRIPTION
**Pull request by superdiaodiao**
## Proposed changes

Issue Number: #31320 

Support two hll functions:

- hll_from_base64
Convert a base64 string(result of function hll_to_base64) into a hll.
- hll_to_base64
Convert an input hll to a base64 string.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by superdiaodiao** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

File: docs/en/docs/sql-manual/sql-functions/hll-functions/hll-from-base64.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+---
+{
+    "title": "HLL_FROM_BASE64",
+    "language": "zh-CN"
```
Comment (HappenLee) [inline docs/en/docs/sql-manual/sql-functions/hll-functions/hll-from-base64.md:4]: en，not `zh-CN`

File: be/src/vec/functions/hll_from_base64.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        auto& null_map = res_null_map->get_data();
+        auto& res = res_data_column->get_data();
+
+        ColumnPtr& argument_column = block.get_by_position(arguments[0]).column;
+        if constexpr (std::is_same_v<typename Impl::ArgumentType, DataTypeString>) {
+            const auto& str_column = static_cast<const ColumnString&>(*argument_column);
+            const ColumnString::Chars& data = str_column.get_chars();
+            const ColumnString::Offsets& offsets = str_column.get_offsets();
+            RETURN_IF_ERROR(Impl::vector(data, offsets, res, null_map, input_rows_count));
+        } else if constexpr (std::is_same_v<typename Impl::ArgumentType, DataTypeArray>) {
```
Comment (HappenLee) [inline be/src/vec/functions/hll_from_base64.cpp:115]: which case the input can be array ?

File: docs/en/docs/sql-manual/sql-functions/hll-functions/hll-from-base64.md (hunk 1 above)
Comment (superdiaodiao) [reply docs/en/docs/sql-manual/sql-functions/hll-functions/hll-from-base64.md:4]: Yeah..........

File: be/src/vec/functions/hll_from_base64.cpp (hunk 2 above)
Comment (superdiaodiao) [reply be/src/vec/functions/hll_from_base64.cpp:115]: Maybe in the future, we could have **hll_from_array** like the function **bitmap_from_array**, so I keep this ability.

File: be/src/vec/functions/hll_from_base64.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            } else if (which_type.is_int64()) {
+                RETURN_IF_ERROR(Impl::template vector<ColumnInt64>(
+                        offset_column_data, nested_column, nested_null_map, res, null_map));
+            } else {
+                return Status::RuntimeError("Illegal column {} of argument of function {}",
+                                            block.get_by_position(arguments[0]).column->get_name(),
+                                            get_name());
+            }
+        } else {
+            return Status::RuntimeError("Illegal column {} of argument of function {}",
```
Comment (HappenLee) [inline be/src/vec/functions/hll_from_base64.cpp:148]: not need to do check

File: be/src/vec/functions/hll_from_base64.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                res.emplace_back();
+                null_map[i] = 1;
+                continue;
+            }
+            curr_decode_buff_len = src_size + 3;
+            if (curr_decode_buff_len > last_decode_buff_len) {
+                decode_buff.resize(curr_decode_buff_len);
+                last_decode_buff_len = curr_decode_buff_len;
+            }
+            int outlen = base64_decode(src_str, src_size, decode_buff.data());
```
Comment (HappenLee) [inline be/src/vec/functions/hll_from_base64.cpp:66]: auto outlen

File: be/src/vec/functions/hll_to_base64.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        for (size_t i = 0; i < size; ++i) {
+            HyperLogLog& hll_val = const_cast<HyperLogLog&>(data[i]);
+
+            cur_ser_size = hll_val.max_serialized_size();
+            if (cur_ser_size > last_ser_size) {
+                last_ser_size = cur_ser_size;
+                ser_buff.resize(cur_ser_size);
+            }
+            hll_val.serialize(reinterpret_cast<uint8_t*>(ser_buff.data()));
+            int outlen = base64_encode((const unsigned char*)ser_buff.data(), cur_ser_size,
```
Comment (HappenLee) [inline be/src/vec/functions/hll_to_base64.cpp:72]: auto outlen

File: be/src/vec/functions/hll_to_base64.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                                       chars_data + encoded_offset);
+            DCHECK(outlen > 0);
+
+            encoded_offset += (int)(4.0 * ceil((double)cur_ser_size / 3.0));
```
Comment (HappenLee) [inline be/src/vec/functions/hll_to_base64.cpp:76]: encoded_offset += outlen;

File: be/src/vec/functions/hll_from_base64.cpp (hunk 4 above)
Comment (superdiaodiao) [reply be/src/vec/functions/hll_from_base64.cpp:66]: ok

File: be/src/vec/functions/hll_from_base64.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    static constexpr auto name = "hll_from_base64";
+
+    static Status vector(const ColumnString::Chars& data, const ColumnString::Offsets& offsets,
+                         std::vector<HyperLogLog>& res, NullMap& null_map,
+                         size_t input_rows_count) {
+        res.reserve(input_rows_count);
+        if (offsets.size() == 0 && input_rows_count == 1) {
+            // For NULL constant
+            res.emplace_back();
+            null_map[0] = 1;
```
[the remark this answers is not in the document]
Comment (superdiaodiao) [reply be/src/vec/functions/hll_from_base64.cpp:47]: yes

File: be/src/vec/functions/hll_to_base64.cpp (hunk 5 above)
Comment (superdiaodiao) [reply be/src/vec/functions/hll_to_base64.cpp:72]: ok

File: be/src/vec/functions/hll_to_base64.cpp (hunk 6 above)
Comment (superdiaodiao) [reply be/src/vec/functions/hll_to_base64.cpp:76]: ok

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run buildall

File: be/src/vec/functions/hll_from_base64.cpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+        res.reserve(input_rows_count);
+
+        std::string decode_buff;
+        int last_decode_buff_len = 0;
+        int curr_decode_buff_len = 0;
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            const char* src_str = reinterpret_cast<const char*>(&data[offsets[i - 1]]);
+            int64_t src_size = offsets[i] - offsets[i - 1];
+            if (0 != src_size % 4) {
```
Comment (HappenLee) [inline be/src/vec/functions/hll_from_base64.cpp:68]: Add comment for const value `4`

File: be/src/vec/functions/hll_from_base64.cpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+                res.emplace_back();
+                null_map[i] = 1;
+                continue;
+            }
+            curr_decode_buff_len = src_size + 3;
```
Comment (HappenLee) [inline be/src/vec/functions/hll_from_base64.cpp:73]: same add comment for the `3` why here 3 ? what meaning?

File: be/src/vec/functions/hll_from_base64.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            }
+            auto outlen = base64_decode(src_str, src_size, decode_buff.data());
+            if (outlen < 0) {
+                res.emplace_back();
+                null_map[i] = 1;
+            } else {
+                doris::Slice decoded_slice(decode_buff.data(), outlen);
+                doris::HyperLogLog hll;
+                if (!hll.deserialize(decoded_slice)) {
+                    return Status::RuntimeError(
```
Comment (HappenLee) [inline be/src/vec/functions/hll_from_base64.cpp:86]: why not set NULL ? or return ERROR?

Comment (superdiaodiao) [reply be/src/vec/functions/hll_from_base64.cpp:86]: I saw bitmap_from_base64() return error, shouldn't we keep the same?

File: be/src/vec/functions/hll_from_base64.cpp (hunk 8 above)
Comment (superdiaodiao) [reply be/src/vec/functions/hll_from_base64.cpp:68]: ok

File: be/src/vec/functions/hll_from_base64.cpp (hunk 9 above)
Comment (superdiaodiao) [reply be/src/vec/functions/hll_from_base64.cpp:73]: ok

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
build feut

**Comment by superdiaodiao** [comment]
run feut

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]

**Pull request merged**